### PR TITLE
Interactive terminal

### PR DESCRIPTION
**Pull request by jmbejar**
Add a terminal section to the git-tutor

- [x] Basic implementation 
- [x] Repeat prompt and last command in output
- [x] Try resize upper sections when the interactive mode is active

### CODE REVIEW

**Comment by jmbejar** [comment]
This PR can be reviewed (it depends on https://github.com/wyeworks/git-tutor/pull/1)

**Comment by rodrigopdl** [comment]
@jmbejar Now that [#1](https://github.com/wyeworks/git-tutor/pull/1) has been `master`, this one can be rebased.

File: lib/git_tutor/gui/interactive_win.rb [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      end
+
+      def execute_command(command)
+        output = begin
+          command.execute
+        rescue Interactive::NotAllowedCommand
+          output = ["The command #{command} is not allowed in this terminal"]
+        end
+
+        output = output[0..@max_output_lines-1]
```
Comment (rodrigopdl) [inline lib/git_tutor/gui/interactive_win.rb]: It looks like this method, `execute_command(command)`, is responsible for two things:

 - Executing the passed in command
 - Drawing the command execution results on the screen


What are your thoughts on splitting these responsibilities apart, ending up with two separate methods:

 - `execute_command(command)`
 - `draw_command_execution_result(output)`

```diff
diff --git a/lib/git_tutor/gui/interactive_win.rb b/lib/git_tutor/gui/interactive_win.rb
index 7407e7a..6b3b0f9 100644
--- a/lib/git_tutor/gui/interactive_win.rb
+++ b/lib/git_tutor/gui/interactive_win.rb
@@ -32,7 +32,8 @@ module GitTutor
             @win.refresh
             break
           else
-            execute_command(command)
+            output = execute_command(command)
+            draw_command_execution_result(output)
           end
         end
       end
@@ -47,12 +48,14 @@ module GitTutor
       end

       def execute_command(command)
-        output = begin
+        begin
           command.execute
         rescue Interactive::NotAllowedCommand
-          output = ["The command #{command} is not allowed in this terminal"]
+          ["The command #{command} is not allowed in this terminal"]
         end
+      end

+      def draw_command_execution_result(output)
         output = output[0..@max_output_lines-1]

         # Remove two extra lines, one for echoing the command
```

File: lib/git_tutor/gui/interactive_win.rb [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        # 1 line command echo
+        # 1 separation between promtp and output lines
+        # 2 lines for prompt and help text
+        @max_output_lines = @window_lines - 7
+
+        @win = Window.new(@window_lines, @window_cols, UPPER_MARGIN_LINES, 1)
+        @win.send(:echo)
+      end
+
+      def run
```
[the remark this answers is not in the document]
Comment (jmbejar) [reply lib/git_tutor/gui/interactive_win.rb]: Done

File: lib/git_tutor/gui/interactive_win.rb [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+      def read_command
+        @win.setpos(@window_lines - 3, 6)
+        @win.send(:curs_set, 1)
+
+        Interactive::Command.new(@win.getstr)
+      end
+
+      def execute_command(command)
+        output = begin
```
[the remark this answers is not in the document]
Comment (jmbejar) [reply lib/git_tutor/gui/interactive_win.rb]: Done

File: lib/git_tutor/gui/interactive_win.rb [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      def clear_line(line_number)
+        @win.setpos(line_number, 0)
+        @win.addstr(' ' * @window_cols)
+      end
+
+      def write(line_number:, text:, bold: false)
+        clear_line(line_number)
+        @win.setpos(line_number, 3)
+
+        font_style = if bold
```
[the remark this answers is not in the document]
Comment (jmbejar) [reply lib/git_tutor/gui/interactive_win.rb]: Fixed

File: lib/git_tutor/gui/interactive_win.rb (hunk 1 above)
Comment (jmbejar) [reply lib/git_tutor/gui/interactive_win.rb]: I tried this idea and it looks good to me. I did some additional changes in this direction (one addition function and more descriptive var names)

**Comment by jmbejar** [comment]
@rodrigopdl  `interactive.rb` is removed now

**Comment by rodrigopdl** [comment]
@jmbejar Thank you!

**Pull request merged**